File: clients/python/test/mut_smoke.py
"""Runtime smoke for the mutating methods, driven against a *disposable copy* of
the `examples/semantic` workspace (set via PAWRLY_WS). Picks the transport from
the env: PAWRLY_GRPC (gRPC) or PAWRLY_REST (REST)."""

import os
import sys
import time

sys.path.insert(0, os.path.join(os.path.dirname(__file__), "..", "src"))

import pawrly  # noqa: E402


def fail(msg: str) -> None:
    print("MUT SMOKE FAIL:", msg)
    sys.exit(1)


ws = os.environ.get("PAWRLY_WS")
if not ws:
    fail("PAWRLY_WS (workspace dir) not set")

if os.environ.get("PAWRLY_GRPC"):
    client = pawrly.PawrlyClient.grpc(os.environ["PAWRLY_GRPC"])
elif os.environ.get("PAWRLY_REST"):
    client = pawrly.PawrlyClient.rest(os.environ["PAWRLY_REST"])
else:
    fail("set PAWRLY_GRPC or PAWRLY_REST")

deadline = time.time() + 10
while True:
    try:
        if client.health().ok:
            break
    except Exception:
        pass
    if time.time() > deadline:
        fail("engine never came up")
    time.sleep(0.1)

print("transport:", client.transport)

# --- reports that always succeed --------------------------------------------
print("reload_config:", client.reload_config())
print("refresh_catalog:", client.refresh_catalog())
print("vacuum_cache:", client.vacuum_cache())

test = client.test_source("data")
print("test_source(data):", test)
if not test.ok:
    fail(f"test_source not ok: {test}")

# --- source lifecycle: add -> list -> remove --------------------------------
info = client.add_source(
    {"name": "extra", "kind": "file", "config": {"path": f"{ws}/data/*.csv"}}
)
print("add_source:", info)
if info.name != "extra":
    fail(f"add_source returned {info}")
if "extra" not in {s.name for s in client.list_sources()}:
    fail("added source not listed")
if client.remove_source("extra") is not True:
    fail("remove_source did not return True")
if "extra" in {s.name for s in client.list_sources()}:
    fail("removed source still listed")
print("source add/remove ok")

# --- materialized lifecycle: materialize -> drop ----------------------------
client.materialize("mut_test", pawrly.MaterializeSpec(sql="SELECT 1 AS n"))
if client.drop_materialized("mut_test") is not True:
    fail("drop_materialized did not return True")
print("materialize/drop ok")

# --- cache-entry ops against a real cache entry -----------------------------
# `customers` has no RLS and a pre-aggregated rollup; this query builds it.
client.semantic_query(
    pawrly.SemanticQuery(
        measures=["customers.customer_total"], dimensions=["customers.region"]
    )
).collect()
entries = client.cache_entries()
if not entries:
    fail("expected a cache entry after the rollup query")
target = f"{entries[0].name.schema}.{entries[0].name.table}"
ro = client.refresh_table(target)
print("refresh_table:", ro)
if ro.rows_written < 0:
    fail("refresh_table bad rows_written")
inv = client.invalidate_cache(target)
print("invalidate_cache:", inv)
if inv is not True:
    fail("invalidate_cache did not return True")

print("MUT SMOKE OK")

File: clients/typescript/test/mut_smoke.mjs
// Runtime smoke for the mutating methods, driven against a *disposable copy* of
// the `examples/semantic` workspace (set via PAWRLY_WS). Picks the transport
// from the env: PAWRLY_GRPC (gRPC) or PAWRLY_REST (REST).
import { PawrlyClient } from "../dist/index.js";

const ws = process.env.PAWRLY_WS;
if (!ws) {
  console.error("PAWRLY_WS (workspace dir) not set");
  process.exit(1);
}

let client;
if (process.env.PAWRLY_GRPC) {
  client = new PawrlyClient({ transport: "grpc", endpoint: process.env.PAWRLY_GRPC });
} else if (process.env.PAWRLY_REST) {
  client = new PawrlyClient({ transport: "rest", baseUrl: process.env.PAWRLY_REST });
} else {
  console.error("set PAWRLY_GRPC or PAWRLY_REST");
  process.exit(1);
}

function fail(msg) {
  console.error("MUT SMOKE FAIL:", msg);
  process.exit(1);
}

const deadline = Date.now() + 10_000;
for (;;) {
  try {
    if ((await client.health()).ok) break;
  } catch {
    /* not up yet */
  }
  if (Date.now() > deadline) fail("engine never came up");
  await new Promise((r) => setTimeout(r, 100));
}

console.log("transport:", client.transport);

// --- reports that always succeed --------------------------------------------
console.log("reloadConfig:", JSON.stringify(await client.reloadConfig()));
console.log("refreshCatalog:", JSON.stringify(await client.refreshCatalog()));
console.log("vacuumCache:", JSON.stringify(await client.vacuumCache()));

const test = await client.testSource("data");
console.log("testSource(data):", JSON.stringify(test));
if (!test.ok) fail(`testSource not ok: ${JSON.stringify(test)}`);

// --- source lifecycle: add -> list -> remove --------------------------------
const info = await client.addSource({ name: "extra", kind: "file", config: { path: `${ws}/data/*.csv` } });
console.log("addSource:", JSON.stringify(info));
if (info.name !== "extra") fail(`addSource returned ${JSON.stringify(info)}`);
if (!(await client.listSources()).some((s) => s.name === "extra")) fail("added source not listed");
if ((await client.removeSource("extra")) !== true) fail("removeSource did not return true");
if ((await client.listSources()).some((s) => s.name === "extra")) fail("removed source still listed");
console.log("source add/remove ok");

// --- materialized lifecycle: materialize -> drop ----------------------------
await client.materialize("mut_test", { kind: "query", sql: "SELECT 1 AS n" });
if ((await client.dropMaterialized("mut_test")) !== true) fail("dropMaterialized did not return true");
console.log("materialize/drop ok");

// --- cache-entry ops against a real cache entry -----------------------------
// `customers` has no RLS and a pre-aggregated rollup; this query builds it.
await (
  await client.semanticQuery({ measures: ["customers.customer_total"], dimensions: ["customers.region"] })
).collect();
const entries = await client.cacheEntries();
if (entries.length === 0) fail("expected a cache entry after the rollup query");
const target = `${entries[0].name.schema}.${entries[0].name.table}`;
const ro = await client.refreshTable(target);
console.log("refreshTable:", JSON.stringify(ro));
if (ro.rowsWritten < 0) fail("refreshTable bad rowsWritten");
const inv = await client.invalidateCache(target);
console.log("invalidateCache:", inv);
if (inv !== true) fail("invalidateCache did not return true");

client.close();
console.log("MUT SMOKE OK");
